### ISSUE TITLE
Parsing ext document missing pages in Kids

### ISSUE THREAD

**Issue by DJWassink** [rkusa/pdfjs#117]
Hello, first cheers for this awesome piece of software.

I am trying to export some pdf's with chrome headless which is working fine until the pdf's get large and chrome crashes. So my idea was to export multiple small pdf's and combine them with this lib.

However I am only getting like 3 pages with the following code:

```js
    const doc = new pdfjs.Document();
    const pdf1 = fs.readFileSync('./test.pdf');
    const ext = new pdfjs.ExternalDocument(pdf1);

    console.info(ext.pageCount);

    doc.addPagesOf(ext);
    doc.asBuffer().then(data => fs.writeFileSync('combined.pdf', data, { encoding: 'binary' }));
```

As you can see I am logging the page count which also says 3, but if I put a logging statement in the `external.js`  on line 13 (either the currently published npm package or master) https://github.com/rkusa/pdfjs/blob/e3cdf6e1060d6974aea5fc5e9f1b6556d2a9d51e/lib/external.js#L12-L14
I get the following output:
![image](https://user-images.githubusercontent.com/1822404/50285149-9e8bae00-045b-11e9-9c08-0c30c0dd19ba.png)

Clearly it detected the 24 pages, but somehow only parsed 3 pages?

If we increase the pdf length *2 (48 pages) we get the double amount, 6 pages:
![image](https://user-images.githubusercontent.com/1822404/50285295-18bc3280-045c-11e9-8f8d-42d7870ae07c.png)

So somehow it does see the amount of pages but the `Kids` are only the first x (as seen by the output).

Am I doing something wrong (like having to await something?) Or does chrome prints some weird pdf?

I tried to dig a bit in the code but have no idea how parsers/lexer work.

I have attached 2 files, the input (test.pdf) and the output (combined.pdf)

Thank in advance.
[combined.pdf](https://github.com/rkusa/pdfjs/files/2698865/combined.pdf)
[test.pdf](https://github.com/rkusa/pdfjs/files/2698866/test.pdf)

**Comment by DJWassink** [comment]
For anyone also having this issue;

Chrome starts after 8 pages splitting the pdf in Kids with each kid having ~8 pages each. Since this lib only sees the Kids as pages we will miss around `amoutOfPage - amountOfPage / 8`. So if chrome starts doing this after 8 pages simply printing with and range of 8 will solve this behavior :).

See the following snippet for an example:

```js
const numOfPages = 50;
const pagesATime = 8;
const doc = new pdfjs.Document();
for (let index = 0; index <= numOfPages; index += pagesATime) {
    const pdfOptions = {
        pageRanges: `${index + 1}-${index + pagesATime}`,
        printBackground: true,
        displayHeaderFooter: false,
        width: '297mm',
        height: '210mm'
    };
    const buf = await page.pdf(pdfOptions);
    const ext = new pdfjs.ExternalDocument(buf);

    doc.addPagesOf(ext);
}
const pdfBuffer = await doc.asBuffer();
fs.writeFileSync('combined.pdf', pdfBuffer, {encoding: 'binary'});
```

The performance is actually pretty decent!
Cheers!

**Comment by rkusa** [comment]
Interesting, seems like the `Kids` array of those PDFs do not contain pages, but other pages objects with their own `Kids` array. So basically a tree structure with leafs being actual pages. This is something I haven't considered for the parser, so definitely a bug that needs to be fixed!
Nice workaround btw! 👍

**Comment by hluz** [comment]
I'm being bitten by this. Trying to use the workaround, but the code snippet does not make sense to me... Considering we have a file buffer to parse as external, not knowing how many pages, how would this work? what/where is the `page.pdf(pdfOptions)` defined?

This is a road blocker for me :-(

**Comment by hluz** [comment]
Ok, I think I understand... the snippet posted by @DJWassink is to ensure that he only creates pdfs with a max of 8 pages. Unfortunately, that does not work for me, since my pdfs that need to be merged are pre-existent and may have arbitrary numbers of pages each. Also, when merging multiple documents, the ones after the one that looses pages don't get included in the combined file at all. Very strange. 

@rkusa, will it help if I provide samples of files that cause the issue?

**Comment by hluz** [comment]
> when merging multiple documents, the ones after the one that looses pages don't get included in the combined file at all

Just to add that this appears to be just a result of the count of number of pages in the merged document is understated, since it is only incremented by the number of kids on the external, although all objects of each kid (which may be multiple pages) are written. Meaning that the parser is loading everything, just that when there are multiple pages in a kid, they are all counted as one page.

**Comment by rkusa** [comment]
Should be fixed on `master`, please let me know if it works correctly now

**Comment by hluz** [comment]
Cool! It is working most of the time! Many thanks in attending to this so quickly.

However, there is one scenario where i'm getting this:

```
I20190214-10:43:15.692(11)? Error: EOL expected but not found
I20190214-10:43:15.692(11)?     at Lexer._error (/Users/hluz/Documents/Development/GreenS/meteor-vue/greenBooks/app/node_modules/pdfjs/lib/parser/lexer.js:128:11)
I20190214-10:43:15.692(11)?     at Lexer.skipEOL (/Users/hluz/Documents/Development/GreenS/meteor-vue/greenBooks/app/node_modules/pdfjs/lib/parser/lexer.js:56:14)
I20190214-10:43:15.692(11)?     at Function.parseLiteral (/Users/hluz/Documents/Development/GreenS/meteor-vue/greenBooks/app/node_modules/pdfjs/lib/object/string.js:114:17)
I20190214-10:43:15.692(11)?     at Function.parse (/Users/hluz/Documents/Development/GreenS/meteor-vue/greenBooks/app/node_modules/pdfjs/lib/object/string.js:27:31)
I20190214-10:43:15.692(11)?     at Object.exports.parse (/Users/hluz/Documents/Development/GreenS/meteor-vue/greenBooks/app/node_modules/pdfjs/lib/object/value.js:20:30)
I20190214-10:43:15.693(11)?     at Function.parse (/Users/hluz/Documents/Development/GreenS/meteor-vue/greenBooks/app/node_modules/pdfjs/lib/object/array.js:33:28)
I20190214-10:43:15.693(11)?     at Object.exports.parse (/Users/hluz/Documents/Development/GreenS/meteor-vue/greenBooks/app/node_modules/pdfjs/lib/object/value.js:20:30)
I20190214-10:43:15.693(11)?     at Function.parseInner (/Users/hluz/Documents/Development/GreenS/meteor-vue/greenBooks/app/node_modules/pdfjs/lib/object/object.js:79:28)
I20190214-10:43:15.693(11)?     at Function.parse (/Users/hluz/Documents/Development/GreenS/meteor-vue/greenBooks/app/node_modules/pdfjs/lib/object/object.js:67:27)
I20190214-10:43:15.693(11)?     at parseObject (/Users/hluz/Documents/Development/GreenS/meteor-vue/greenBooks/app/node_modules/pdfjs/lib/object/reference.js:128:22)
I20190214-10:43:15.693(11)?     at PDFReference.get [as object] (/Users/hluz/Documents/Development/GreenS/meteor-vue/greenBooks/app/node_modules/pdfjs/lib/object/reference.js:15:17)
I20190214-10:43:15.693(11)?     at Function.addObjectsRecursive (/Users/hluz/Documents/Development/GreenS/meteor-vue/greenBooks/app/node_modules/pdfjs/lib/parser/parser.js:61:35)
I20190214-10:43:15.694(11)?     at Function.addObjectsRecursive (/Users/hluz/Documents/Development/GreenS/meteor-vue/greenBooks/app/node_modules/pdfjs/lib/parser/parser.js:77:18)
I20190214-10:43:15.694(11)?     at Function.addObjectsRecursive (/Users/hluz/Documents/Development/GreenS/meteor-vue/greenBooks/app/node_modules/pdfjs/lib/parser/parser.js:68:16)
I20190214-10:43:15.694(11)?     at Function.addObjectsRecursive (/Users/hluz/Documents/Development/GreenS/meteor-vue/greenBooks/app/node_modules/pdfjs/lib/parser/parser.js:65:16)
I20190214-10:43:15.694(11)?     at Function.addObjectsRecursive (/Users/hluz/Documents/Development/GreenS/meteor-vue/greenBooks/app/node_modules/pdfjs/lib/parser/parser.js:77:18)
```
In case that puts you in the right track, I got that when I tried to merge a previously merged pdf.

**Comment by rkusa** [comment]
@hluz is this a non-confidential PDF you could share? (per mail would be fine as well, see my github profile)?

**Comment by hluz** [comment]
> per mail would be fine as well, see my github profile

Sent.

**Comment by rkusa** [comment]
Thanks for the PDF, should be fixed

**Comment by hluz** [comment]
Tested. Perfect now. Thanks!

**Comment by tomermoshe** [comment]
Can you please publish a new version to npm with this fix? The latest version on npm is 2.1.0 that was released 6 month ago. Thanks!

**Comment by rkusa** [comment]
@tomermoshe valid request, published as `2.2.0`
